src/v/cluster/partition_balancer_planner.cc

```diff
@@ -358,6 +358,105 @@ void partition_balancer_planner::get_unavailable_nodes_reassignments(
     }
 }
 
+/// Try to fix ntps that have several replicas in one rack (these ntps can
+/// appear because rack awareness constraint is not a hard constraint, e.g. when
+/// a rack dies and we move all replicas that resided on dead nodes to live
+/// ones).
+///
+/// We go over all such ntps (a list maintained by partition_balancer_state) and
+/// if the number of currently live racks is more than the number of racks that
+/// the ntp is replicated to, we try to schedule a move. For each rack we
+/// arbitrarily choose the first appearing replica to remain there (note: this
+/// is probably not optimal choice).
+void partition_balancer_planner::get_rack_constraint_repair_reassignments(
+  plan_data& result, reallocation_request_state& rrs) {
+    if (_state.ntps_with_broken_rack_constraint().empty()) {
```

Comment:
nit: this condition is already check by the caller

Reply:
I don't think this is true - we can end here e.g. if there are some unavailable nodes but no violating ntps. Although it might make sense to make it so! Would be easier to read

Reply:
Reshuffled the main function a bit, not sure if this is much cleaner, but the idea is that a planner pass should decide itself if it needs to run or not, but we also want to avoid loading ntp sizes in the happy case (this is why we need an early exit if there are no violations).